File: backend/app/notifications/email.py
"""Enviament d'avisos per correu."""

from __future__ import annotations

import logging
import smtplib
from email.message import EmailMessage

from jinja2 import Template

from app.config import settings
from app.core.time import to_local
from app.models import Alert
from app.models.enums import AlertSeverity

logger = logging.getLogger(__name__)

SEVERITY_LABEL = {
    AlertSeverity.CRITICAL: "Urgent",
    AlertSeverity.WARNING: "Atencio",
    AlertSeverity.INFO: "Informatiu",
}
SEVERITY_COLOR = {
    AlertSeverity.CRITICAL: "#dc2626",
    AlertSeverity.WARNING: "#d97706",
    AlertSeverity.INFO: "#2563eb",
}

DIGEST_TEMPLATE = Template(
    """<!doctype html>
<html lang="ca">
  <body style="font-family: -apple-system, Segoe UI, Roboto, sans-serif; color:#0f172a;">
    <h2 style="margin-bottom:4px;">{{ title }}</h2>
    <p style="color:#64748b; margin-top:0;">{{ subtitle }}</p>
    {% for alert in alerts %}
      <div style="border-left:4px solid {{ colors[alert.severity] }};
                  background:#f8fafc; padding:12px 16px; margin:12px 0;">
        <div style="font-size:12px; text-transform:uppercase; letter-spacing:.05em;
                    color:{{ colors[alert.severity] }};">
          {{ labels[alert.severity] }}
          {%- if alert.ledger_name %} · {{ alert.ledger_name }}{% endif %}
        </div>
        <div style="font-weight:600; margin:4px 0;">{{ alert.title }}</div>
        <div style="color:#334155;">{{ alert.body }}</div>
        <div style="color:#94a3b8; font-size:12px; margin-top:6px;">{{ alert.created }}</div>
      </div>
    {% endfor %}
    <p style="margin-top:24px;">
      <a href="{{ base_url }}" style="color:#2563eb;">Obre la comptabilitat</a>
    </p>
  </body>
</html>"""
)


def render_digest(alerts: list[Alert], title: str, subtitle: str) -> tuple[str, str]:
    """Retorna (html, text) del resum d'avisos."""
    entries = [
        {
            "severity": alert.severity,
            "title": alert.title,
            "body": alert.body,
            "ledger_name": alert.ledger.name if getattr(alert, "ledger", None) else "",
            "created": to_local(alert.created_at).strftime("%d/%m/%Y %H:%M"),
        }
        for alert in alerts
    ]
    html = DIGEST_TEMPLATE.render(
        alerts=entries,
        title=title,
        subtitle=subtitle,
        labels=SEVERITY_LABEL,
        colors=SEVERITY_COLOR,
        base_url=settings.public_base_url,
    )
    lines = [title, subtitle, ""]
    for entry in entries:
        lines.append(f"[{SEVERITY_LABEL[entry['severity']]}] {entry['title']}")
        if entry["body"]:
            lines.append(f"  {entry['body']}")
        lines.append("")
    lines.append(settings.public_base_url)
    return html, "\n".join(lines)


def send_email(subject: str, html: str, text: str, recipients: list[str] | None = None) -> bool:
    """Envia un correu. Retorna si s'ha pogut enviar."""
    targets = recipients or settings.alert_recipients
    if not settings.smtp_configured or not targets:
        logger.info("Correu no configurat: no s'envia «%s»", subject)
        return False

    message = EmailMessage()
    message["Subject"] = subject
    message["From"] = settings.smtp_from
    message["To"] = ", ".join(targets)
    message.set_content(text)
    message.add_alternative(html, subtype="html")

    try:
        if settings.smtp_port == 465:
            with smtplib.SMTP_SSL(settings.smtp_host, settings.smtp_port, timeout=30) as server:
                _login_and_send(server, message)
        else:
            with smtplib.SMTP(settings.smtp_host, settings.smtp_port, timeout=30) as server:
                if settings.smtp_starttls:
                    server.starttls()
                _login_and_send(server, message)
    except (smtplib.SMTPException, OSError) as exc:
        logger.error("No s'ha pogut enviar el correu «%s»: %s", subject, exc)
        return False

    logger.info("Correu enviat: %s → %s", subject, targets)
    return True


def _login_and_send(server: smtplib.SMTP, message: EmailMessage) -> None:
    if settings.smtp_user:
        server.login(settings.smtp_user, settings.smtp_password)
    server.send_message(message)

File: src/lib/email.ts
/**
 * Sending alerts by mail.
 *
 * Alert text carries merchant names that come from the bank, so the HTML body
 * is built with Hono's `html` tag, which escapes whatever is interpolated into
 * it. The mail itself is Catalan: it is read by the people using the app.
 *
 * Translated from `backend/app/notifications/email.py`.
 */

import { html } from "hono/html";
import nodemailer from "nodemailer";

import type { AlertSeverity } from "../db/schema/enums.ts";
import { config, smtpConfigured } from "./config.ts";

const SEVERITY_LABEL: Record<AlertSeverity, string> = {
  critical: "Urgent",
  warning: "Atencio",
  info: "Informatiu",
};

const SEVERITY_COLOR: Record<AlertSeverity, string> = {
  critical: "#dc2626",
  warning: "#d97706",
  info: "#2563eb",
};

/** One alert, with its workspace name and an already-formatted date. */
export interface SummaryEntry {
  severity: AlertSeverity;
  title: string;
  body: string;
  ledgerName: string;
  created: string;
}

export interface Summary {
  html: string;
  text: string;
}

/** Returns the HTML body and the plain-text body of the alert summary. */
export async function renderSummary(
  entries: readonly SummaryEntry[],
  title: string,
  subtitle: string,
): Promise<Summary> {
  const blocks = await Promise.all(
    entries.map(
      (e) => html`
        <div
          style="border-left:4px solid ${SEVERITY_COLOR[e.severity]};
                 background:#f8fafc; padding:12px 16px; margin:12px 0;"
        >
          <div
            style="font-size:12px; text-transform:uppercase; letter-spacing:.05em;
                   color:${SEVERITY_COLOR[e.severity]};"
          >
            ${SEVERITY_LABEL[e.severity]}${e.ledgerName !== "" ? ` · ${e.ledgerName}` : ""}
          </div>
          <div style="font-weight:600; margin:4px 0;">${e.title}</div>
          <div style="color:#334155;">${e.body}</div>
          <div style="color:#94a3b8; font-size:12px; margin-top:6px;">${e.created}</div>
        </div>
      `,
    ),
  );

  const document = await html`<!doctype html>
    <html lang="ca">
      <body style="font-family: -apple-system, Segoe UI, Roboto, sans-serif; color:#0f172a;">
        <h2 style="margin-bottom:4px;">${title}</h2>
        <p style="color:#64748b; margin-top:0;">${subtitle}</p>
        ${blocks}
        <p style="margin-top:24px;">
          <a href="${config.publicBaseUrl}" style="color:#2563eb;">Obre la comptabilitat</a>
        </p>
      </body>
    </html>`;

  const lines: string[] = [title, subtitle, ""];
  for (const entry of entries) {
    lines.push(`[${SEVERITY_LABEL[entry.severity]}] ${entry.title}`);
    if (entry.body !== "") lines.push(`  ${entry.body}`);
    lines.push("");
  }
  lines.push(config.publicBaseUrl);

  return { html: String(document), text: lines.join("\n") };
}

/**
 * Sends a mail. Returns whether it went out.
 *
 * NOTE: `smtpConfigured` requires `ALERT_RECIPIENTS` to be non-empty, even
 * though a workspace can have recipients of its own. That is the existing
 * behaviour; it means that without a general list, per-workspace alerts do not
 * go out either. Kept as it was, so nothing changes silently.
 */
export async function sendMail(
  subject: string,
  htmlBody: string,
  textBody: string,
  recipients?: readonly string[],
): Promise<boolean> {
  const target = recipients ?? config.alertRecipients;
  if (!smtpConfigured() || target.length === 0) {
    console.info(`[correu] no configurat: no s'envia «${subject}»`);
    return false;
  }

  const transport = nodemailer.createTransport({
    host: config.smtpHost,
    port: config.smtpPort,
    // 465 is encrypted from the first byte; the rest start in the clear and
    // upgrade to TLS with STARTTLS, as Python's `smtplib` did.
    secure: config.smtpPort === 465,
    requireTLS: config.smtpPort !== 465 && config.smtpStarttls,
    ...(config.smtpUser !== ""
      ? { auth: { user: config.smtpUser, pass: config.smtpPassword } }
      : {}),
    connectionTimeout: 30_000,
    greetingTimeout: 30_000,
    socketTimeout: 30_000,
  });

  try {
    await transport.sendMail({
      from: config.smtpFrom,
      to: [...target],
      subject,
      text: textBody,
      html: htmlBody,
    });
  } catch (error) {
    // On some servers the error message carries the SMTP password: only the
    // short text is logged, never the whole object.
    const detail = error instanceof Error ? error.message : String(error);
    console.error(`[correu] no s'ha pogut enviar «${subject}»: ${detail}`);
    return false;
  } finally {
    transport.close();
  }

  console.info(`[correu] enviat: ${subject} → ${target.join(", ")}`);
  return true;
}
